File: cli/lib/dobase/commands/todos.rb
# frozen_string_literal: true

module Dobase
  module Commands
    class Todos < Command
      REPEATS = %w[daily weekly monthly].freeze
      ITEM_FLAGS = {
        description: [ "TEXT", "Description (plain text, or HTML with --html)" ],
        html: [ nil, "The description is HTML" ],
        due: [ "DATE", "Due date: YYYY-MM-DD, today, tomorrow or none" ],
        assignee: [ "USER", "me, none, a user id, email or name" ],
        repeat: [ "RULE", "#{REPEATS.join(", ")} or none" ]
      }.freeze

      noun "todo", "Todos on lists (todos tools)"
      noun "todolist", "Lists in a todos tool"

      command "todo list", "Show a todos tool: its lists with their open and recently completed todos", args: %w[TOOL],
        flags: { completed: [ nil, "Show every completed todo instead" ] } do |ref, completed: false|
        todos_tool = tool(ref, "todos")
        todo = get("/tools/#{todos_tool["id"]}/todo", completed: completed || nil)

        output(todo) do
          say "#{todos_tool["name"]} (todos #{todos_tool["id"]}) #{todo["url"]}"
          todo["lists"].each do |list|
            say
            say "#{list["title"]} [list #{list["id"]}]"
            say "  (no #{"completed " if completed}todos)" if list["items"].empty?
            table(list["items"].map { |item|
              [ "#{todos_tool["id"]}/#{item["id"]}", item["completed"] ? "[x]" : "[ ]", item["title"], item_summary(item) ]
            })
          end
        end
      end

      command "todo show", "Show a todo with its description, comments and attachments", args: %w[TOOL/ITEM] do |ref|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        item = get("/tools/#{todos_tool["id"]}/todo/items/#{id}")

        output(item) do
          say "#{item["title"]} (todo #{todos_tool["id"]}/#{item["id"]})"
          field "List", "#{todos_tool["name"]} › #{item.dig("list", "title")}"
          field "Status", item["completed"] ? "done #{moment(item["completed_at"])}" : "open"
          field "Assignee", person(item["assignee"])
          field "Due", item["due_date"]
          field "Repeats", item["recurrence_rule"]
          field "Created", [ moment(item["created_at"]), item.dig("creator", "name") ].compact.join(" by ")
          field "URL", item["url"]

          unless item["description"].strip.empty?
            say
            say "Description:"
            paragraph item["description"]
          end

          say
          say "Comments (#{item["comments"].size}):"
          item["comments"].each do |comment|
            say "  #{comment.dig("user", "name")} · #{moment(comment["created_at"])} [comment #{comment["id"]}]"
            paragraph comment["body"], indent: 4
          end

          unless item["attachments"].empty?
            say
            say "Attachments (#{item["attachments"].size}):"
            table(item["attachments"].map { |attachment| [ attachment["filename"], bytes(attachment["file_size"]), attachment["download_url"] ] })
          end
        end
      end

      command "todo create", "Add a todo to the bottom of a list (the first list unless --list)", args: %w[TOOL TITLE],
        flags: ITEM_FLAGS.merge(list: [ "LIST", "List id or name" ]) do |ref, title, list: nil, **options|
        todos_tool = tool(ref, "todos")
        target = find_list(todos_tool, list)

        item = post("/todo_lists/#{target["id"]}/items", item: item_attributes(todos_tool, options).merge(title: title))
        output(item) { say "Created todo #{todos_tool["id"]}/#{item["id"]} #{quoted(item["title"])} in #{target["title"]}: #{item["url"]}" }
      end

      command "todo update", "Change a todo's title, description, due date, assignee or repeat", args: %w[TOOL/ITEM],
        flags: ITEM_FLAGS.merge(title: [ "TEXT", "New title" ]) do |ref, title: nil, **options|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        attributes = item_attributes(todos_tool, options)
        attributes[:title] = title if title
        raise UsageError, "Nothing to update. See `dobase help todo`." if attributes.empty?

        item = patch("/tools/#{todos_tool["id"]}/todo/items/#{id}", item: attributes)
        output(item) { say "Updated todo #{todos_tool["id"]}/#{item["id"]} #{quoted(item["title"])}." }
      end

      command "todo finish", "Mark a todo as done (a repeating todo comes back with its next due date)", args: %w[TOOL/ITEM] do |ref|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        item = post("/tools/#{todos_tool["id"]}/todo/items/#{id}/completion")

        output(item) do
          say "Completed todo #{todos_tool["id"]}/#{item["id"]} #{quoted(item["title"])}."
          say "It repeats #{item["recurrence_rule"]}, so the next one is on the list." if item["recurrence_rule"]
        end
      end

      command "todo reopen", "Mark a completed todo as not done", args: %w[TOOL/ITEM] do |ref|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        item = delete("/tools/#{todos_tool["id"]}/todo/items/#{id}/completion")
        output(item) { say "Reopened todo #{todos_tool["id"]}/#{item["id"]} #{quoted(item["title"])}." }
      end

      command "todo move", "Move a todo to another list, or to a position within its list", args: %w[TOOL/ITEM [LIST]],
        flags: { position: [ "N", "Position in the list, 1 = top (default: bottom)" ] } do |ref, list = nil, position: nil|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        raise UsageError, "Give a LIST, a --position, or both." if list.nil? && position.nil?

        target = find_list(todos_tool, list) if list
        body = { todo_list_id: target&.dig("id"), position: position && [ position.to_i - 1, 0 ].max }.compact
        item = patch("/tools/#{todos_tool["id"]}/todo/items/#{id}/position", body)
        output(item) { say "Moved todo #{todos_tool["id"]}/#{item["id"]} #{quoted(item["title"])} to #{item.dig("list", "title")}, position #{item["position"] + 1}." }
      end

      command "todo delete", "Delete a todo permanently, with its comments and attachments", args: %w[TOOL/ITEM] do |ref|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        delete("/tools/#{todos_tool["id"]}/todo/items/#{id}")
        output(nil) { say "Deleted todo #{todos_tool["id"]}/#{id}." }
      end

      command "todo comment", "Comment on a todo", args: %w[TOOL/ITEM TEXT],
        flags: { html: [ nil, "TEXT is HTML" ] } do |ref, body, html: false|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        comment = post("/tools/#{todos_tool["id"]}/todo/items/#{id}/comments", body: rich_text(body, html: html))
        output(comment) { say "Commented on todo #{todos_tool["id"]}/#{id} [comment #{comment["id"]}]." }
      end

      command "todo attach", "Attach files to a todo (25 MB max each)", args: %w[TOOL/ITEM PATH...] do |ref, *paths|
        todos_tool, id = tool_and_id(ref, "todos", "item")
        missing = paths.reject { |path| File.file?(path) }
        raise UsageError, "No such file: #{missing.join(", ")}" if missing.any?

        attachments = paths.map { |path| client.upload("/tools/#{todos_tool["id"]}/todo/items/#{id}/attachments", files: { file: path }) }
        output(attachments) { attachments.each { |attachment| say "Attached #{attachment["filename"]} (#{bytes(attachment["file_size"])}) to todo #{todos_tool["id"]}/#{id}." } }
      end

      command "todolist create", "Add a list to the end of a todos tool", args: %w[TOOL TITLE] do |ref, title|
        todos_tool = tool(ref, "todos")
        list = post("/tools/#{todos_tool["id"]}/todo/lists", title: title)
        output(list) { say "Created list #{todos_tool["id"]}/#{list["id"]} #{quoted(list["title"])}." }
      end

      command "todolist rename", "Rename a list", args: %w[TOOL/LIST TITLE] do |ref, title|
        todos_tool, id = tool_and_id(ref, "todos", "list")
        list = patch("/tools/#{todos_tool["id"]}/todo/lists/#{id}", title: title)
        output(list) { say "Renamed list #{todos_tool["id"]}/#{list["id"]} to #{quoted(list["title"])}." }
      end

      command "todolist delete", "Delete a list and every todo on it", args: %w[TOOL/LIST] do |ref|
        todos_tool, id = tool_and_id(ref, "todos", "list")
        delete("/tools/#{todos_tool["id"]}/todo/lists/#{id}")
        output(nil) { say "Deleted list #{todos_tool["id"]}/#{id}." }
      end

      private

      def item_summary(item)
        [
          ("done #{day(item["completed_at"])}" if item["completed"]),
          ("due #{item["due_date"]}" if item["due_date"]),
          ("@#{item.dig("assignee", "name")}" if item["assignee"]),
          ("repeats #{item["recurrence_rule"]}" if item["recurrence_rule"]),
          (count(item["comments_count"], "comment") if item["comments_count"].positive?),
          (count(item["attachments_count"], "file") if item["attachments_count"].positive?)
        ].compact.join("  ")
      end

      def find_list(todos_tool, ref)
        lists = get("/tools/#{todos_tool["id"]}/todo")["lists"]
        return lists.first || raise(Error, "#{todos_tool["name"]} has no lists.") if ref.nil?

        match = lists.find { |list| list["id"].to_s == ref } ||
          lists.find { |list| list["title"].casecmp?(ref) } ||
          lists.find { |list| list["title"].downcase.start_with?(ref.downcase) }
        match || raise(Error, "No list matches #{quoted(ref)}. Lists: #{lists.map { |list| list["title"] }.join(", ")}")
      end

      def item_attributes(todos_tool, options)
        attributes = {}
        attributes[:description] = rich_text(options[:description], html: options[:html]) if options[:description]
        attributes[:due_date] = date_param(options[:due]) if options[:due]
        attributes[:assigned_user_id] = user_id(todos_tool, options[:assignee]) if options[:assignee]
        if options[:repeat]
          raise UsageError, "--repeat must be one of: #{REPEATS.join(", ")}, none" unless (REPEATS + [ "none" ]).include?(options[:repeat])
          attributes[:recurrence_rule] = options[:repeat] == "none" ? nil : options[:repeat]
        end
        attributes
      end
    end
  end
end

File: cli/src/commands/todos.rs
use serde_json::{Map, Value, json};

use super::boards::{attach_files, find_named, show_details, zero_based};
use crate::command::{Args, Ctx, Definition, Flag, Result, command, count, date_param, day, flag, moment, person, quoted, switch, usage};
use crate::value::{Json, compact, join};

const REPEATS: [&str; 3] = ["daily", "weekly", "monthly"];

fn item_flags() -> Vec<Flag> {
    vec![
        flag("description", "TEXT", "Description (plain text, or HTML with --html)"),
        switch("html", "The description is HTML"),
        flag("due", "DATE", "Due date: YYYY-MM-DD, today, tomorrow or none"),
        flag("assignee", "USER", "me, none, a user id, email or name"),
        flag("repeat", "RULE", format!("{} or none", REPEATS.join(", "))),
    ]
}

pub fn definitions() -> Vec<Definition> {
    vec![
        command(
            "todo list",
            "Show a todos tool: its lists with their open and recently completed todos",
            &["TOOL"],
            vec![switch("completed", "Show every completed todo instead")],
            list,
        ),
        command("todo show", "Show a todo with its description, comments and attachments", &["TOOL/ITEM"], vec![], show),
        command(
            "todo create",
            "Add a todo to the bottom of a list (the first list unless --list)",
            &["TOOL", "TITLE"],
            [item_flags(), vec![flag("list", "LIST", "List id or name")]].concat(),
            create,
        ),
        command(
            "todo update",
            "Change a todo's title, description, due date, assignee or repeat",
            &["TOOL/ITEM"],
            [item_flags(), vec![flag("title", "TEXT", "New title")]].concat(),
            update,
        ),
        command("todo finish", "Mark a todo as done (a repeating todo comes back with its next due date)", &["TOOL/ITEM"], vec![], finish),
        command("todo reopen", "Mark a completed todo as not done", &["TOOL/ITEM"], vec![], reopen),
        command(
            "todo move",
            "Move a todo to another list, or to a position within its list",
            &["TOOL/ITEM", "[LIST]"],
            vec![flag("position", "N", "Position in the list, 1 = top (default: bottom)")],
            move_item,
        ),
        command("todo delete", "Delete a todo permanently, with its comments and attachments", &["TOOL/ITEM"], vec![], delete),
        command("todo comment", "Comment on a todo", &["TOOL/ITEM", "TEXT"], vec![switch("html", "TEXT is HTML")], comment),
        command("todo attach", "Attach files to a todo (25 MB max each)", &["TOOL/ITEM", "PATH..."], vec![], attach),
        command("todolist create", "Add a list to the end of a todos tool", &["TOOL", "TITLE"], vec![], list_create),
        command("todolist rename", "Rename a list", &["TOOL/LIST", "TITLE"], vec![], list_rename),
        command("todolist delete", "Delete a list and every todo on it", &["TOOL/LIST"], vec![], list_delete),
    ]
}

fn list(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let completed = args.on("completed");
    let tool = ctx.tool(args.at(0), Some("todos"))?;
    let todo = ctx.get(&format!("/tools/{}/todo", tool["id"].s()), &[("completed", completed.then(|| "true".to_string()))])?;

    ctx.output(&todo, |ctx| {
        ctx.say(format!("{} (todos {}) {}", tool["name"].s(), tool["id"].s(), todo["url"].s()));
        for list in todo["lists"].items() {
            ctx.blank();
            ctx.say(format!("{} [list {}]", list["title"].s(), list["id"].s()));
            let items = list["items"].items();
            if items.is_empty() {
                ctx.say(format!("  (no {}todos)", if completed { "completed " } else { "" }));
            }
            let rows = items
                .iter()
                .map(|item| {
                    vec![
                        format!("{}/{}", tool["id"].s(), item["id"].s()),
                        if item["completed"].truthy() { "[x]" } else { "[ ]" }.to_string(),
                        item["title"].s(),
                        item_summary(item),
                    ]
                })
                .collect();
            ctx.table(rows, 2);
        }
        Ok(())
    })
}

fn show(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    let item = ctx.get(&format!("/tools/{}/todo/items/{id}", tool["id"].s()), &[])?;

    ctx.output(&item, |ctx| {
        ctx.say(format!("{} (todo {}/{})", item["title"].s(), tool["id"].s(), item["id"].s()));
        ctx.field("List", format!("{} › {}", tool["name"].s(), item["list"]["title"].s()));
        let status = if item["completed"].truthy() {
            format!("done {}", moment(&item["completed_at"]).unwrap_or_default())
        } else {
            "open".to_string()
        };
        ctx.field("Status", status);
        ctx.field("Assignee", person(&item["assignee"]));
        ctx.field("Due", item["due_date"].s());
        ctx.field("Repeats", item["recurrence_rule"].s());
        ctx.field("Created", join([moment(&item["created_at"]), item["creator"]["name"].opt()], " by "));
        ctx.field("URL", item["url"].s());
        show_details(ctx, &item);
        Ok(())
    })
}

fn create(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let tool = ctx.tool(args.at(0), Some("todos"))?;
    let target = find_list(ctx, &tool, args.flag("list"))?;

    let mut attributes = item_attributes(ctx, &tool, args)?;
    attributes.insert("title".into(), json!(args.at(1)));
    let item = ctx.post(&format!("/todo_lists/{}/items", target["id"].s()), json!({ "item": attributes }))?;
    ctx.output(&item, |ctx| {
        ctx.say(format!(
            "Created todo {}/{} {} in {}: {}",
            tool["id"].s(),
            item["id"].s(),
            quoted(&item["title"].s()),
            target["title"].s(),
            item["url"].s()
        ));
        Ok(())
    })
}

fn update(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    let mut attributes = item_attributes(ctx, &tool, args)?;
    if let Some(title) = args.flag("title") {
        attributes.insert("title".into(), json!(title));
    }
    if attributes.is_empty() {
        usage!("Nothing to update. See `dobase help todo`.");
    }

    let item = ctx.patch(&format!("/tools/{}/todo/items/{id}", tool["id"].s()), json!({ "item": attributes }))?;
    ctx.output(&item, |ctx| {
        ctx.say(format!("Updated todo {}/{} {}.", tool["id"].s(), item["id"].s(), quoted(&item["title"].s())));
        Ok(())
    })
}

fn finish(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    let item = ctx.post(&format!("/tools/{}/todo/items/{id}/completion", tool["id"].s()), json!({}))?;

    ctx.output(&item, |ctx| {
        ctx.say(format!("Completed todo {}/{} {}.", tool["id"].s(), item["id"].s(), quoted(&item["title"].s())));
        if let Some(rule) = item["recurrence_rule"].opt() {
            ctx.say(format!("It repeats {rule}, so the next one is on the list."));
        }
        Ok(())
    })
}

fn reopen(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    let item = ctx.delete(&format!("/tools/{}/todo/items/{id}/completion", tool["id"].s()))?;
    ctx.output(&item, |ctx| {
        ctx.say(format!("Reopened todo {}/{} {}.", tool["id"].s(), item["id"].s(), quoted(&item["title"].s())));
        Ok(())
    })
}

fn move_item(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    let (list, position) = (args.get(1), args.flag("position"));
    if list.is_none() && position.is_none() {
        usage!("Give a LIST, a --position, or both.");
    }

    let target = match list {
        Some(list) => find_list(ctx, &tool, Some(list))?["id"].clone(),
        None => Value::Null,
    };
    let body = compact(json!({ "todo_list_id": target, "position": position.map(zero_based) }));
    let item = ctx.patch(&format!("/tools/{}/todo/items/{id}/position", tool["id"].s()), body)?;
    ctx.output(&item, |ctx| {
        ctx.say(format!(
            "Moved todo {}/{} {} to {}, position {}.",
            tool["id"].s(),
            item["id"].s(),
            quoted(&item["title"].s()),
            item["list"]["title"].s(),
            item["position"].int() + 1
        ));
        Ok(())
    })
}

fn delete(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    ctx.delete(&format!("/tools/{}/todo/items/{id}", tool["id"].s()))?;
    ctx.output(&Value::Null, |ctx| {
        ctx.say(format!("Deleted todo {}/{id}.", tool["id"].s()));
        Ok(())
    })
}

fn comment(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    let body = ctx.rich_text(args.at(1), args.on("html"))?;
    let comment = ctx.post(&format!("/tools/{}/todo/items/{id}/comments", tool["id"].s()), json!({ "body": body }))?;
    ctx.output(&comment, |ctx| {
        ctx.say(format!("Commented on todo {}/{id} [comment {}].", tool["id"].s(), comment["id"].s()));
        Ok(())
    })
}

fn attach(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "item")?;
    let path = format!("/tools/{}/todo/items/{id}/attachments", tool["id"].s());
    let label = format!("todo {}/{id}", tool["id"].s());
    attach_files(ctx, &path, args.rest(1), &label)
}

fn list_create(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let tool = ctx.tool(args.at(0), Some("todos"))?;
    let list = ctx.post(&format!("/tools/{}/todo/lists", tool["id"].s()), json!({ "title": args.at(1) }))?;
    ctx.output(&list, |ctx| {
        ctx.say(format!("Created list {}/{} {}.", tool["id"].s(), list["id"].s(), quoted(&list["title"].s())));
        Ok(())
    })
}

fn list_rename(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "list")?;
    let list = ctx.patch(&format!("/tools/{}/todo/lists/{id}", tool["id"].s()), json!({ "title": args.at(1) }))?;
    ctx.output(&list, |ctx| {
        ctx.say(format!("Renamed list {}/{} to {}.", tool["id"].s(), list["id"].s(), quoted(&list["title"].s())));
        Ok(())
    })
}

fn list_delete(ctx: &mut Ctx, args: &Args) -> Result<()> {
    let (tool, id) = ctx.tool_and_id(args.at(0), "todos", "list")?;
    ctx.delete(&format!("/tools/{}/todo/lists/{id}", tool["id"].s()))?;
    ctx.output(&Value::Null, |ctx| {
        ctx.say(format!("Deleted list {}/{id}.", tool["id"].s()));
        Ok(())
    })
}

fn item_summary(item: &Value) -> String {
    join(
        [
            item["completed"].truthy().then(|| format!("done {}", day(&item["completed_at"]).unwrap_or_default())),
            item["due_date"].opt().map(|due| format!("due {due}")),
            item["assignee"].truthy().then(|| format!("@{}", item["assignee"]["name"].s())),
            item["recurrence_rule"].opt().map(|rule| format!("repeats {rule}")),
            (item["comments_count"].int() > 0).then(|| count(item["comments_count"].int(), "comment")),
            (item["attachments_count"].int() > 0).then(|| count(item["attachments_count"].int(), "file")),
        ],
        "  ",
    )
}

fn find_list(ctx: &mut Ctx, tool: &Value, reference: Option<&str>) -> Result<Value> {
    let todo = ctx.get(&format!("/tools/{}/todo", tool["id"].s()), &[])?;
    find_named(todo["lists"].items(), reference, "title", "list", "Lists", &tool["name"].s())
}

fn item_attributes(ctx: &mut Ctx, tool: &Value, args: &Args) -> Result<Map<String, Value>> {
    let mut attributes = Map::new();
    if let Some(description) = args.flag("description") {
        attributes.insert("description".into(), json!(ctx.rich_text(description, args.on("html"))?));
    }
    if let Some(due) = args.flag("due") {
        attributes.insert("due_date".into(), json!(date_param(due)?));
    }
    if let Some(assignee) = args.flag("assignee") {
        attributes.insert("assigned_user_id".into(), ctx.user_id(tool, assignee)?);
    }
    if let Some(repeat) = args.flag("repeat") {
        if !REPEATS.contains(&repeat) && repeat != "none" {
            usage!("--repeat must be one of: {}, none", REPEATS.join(", "));
        }
        attributes.insert("recurrence_rule".into(), if repeat == "none" { Value::Null } else { json!(repeat) });
    }
    Ok(attributes)
}
